src/main/java/org/logstash/beats/BeatsHandler.java

```diff
@@ -0,0 +1,104 @@
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.LogManager;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@ChannelHandler.Sharable
+public class BeatsHandler extends SimpleChannelInboundHandler<Batch> {
+    private static Logger logger = LogManager.getLogger(BeatsHandler.class.getName());
+    private final AtomicBoolean processing = new AtomicBoolean(false);
+    private final IMessageListener messageListener;
+    private ChannelHandlerContext context;
+
+
+    public BeatsHandler(IMessageListener listener) {
+        messageListener = listener;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        context = ctx;
+        messageListener.onNewConnection(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        messageListener.onConnectionClose(ctx);
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, Batch batch) {
+        logger.debug("Received a new payload");
+
+        processing.compareAndSet(false, true);
+
+        for(Message message : batch.getMessages()) {
+            logger.debug("Sending a new message for the listener, sequence: {}", message.getSequence());
+            messageListener.onNewMessage(ctx, message);
+
+            if(needAck(message)) {
+                ack(ctx, message);
+            }
+        }
+
+        processing.compareAndSet(true, false);
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        messageListener.onException(ctx);
+        logger.error("Exception", cause);
+        ctx.close();
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object event) {
+        if(event instanceof IdleStateEvent) {
+            IdleStateEvent e = (IdleStateEvent) event;
+
+            if(e.state() == IdleState.WRITER_IDLE) {
+                sendKeepAlive();
+            } else if(e.state() == IdleState.READER_IDLE) {
+                clientTimeout();
+            }
+        }
+    }
+
+    private boolean needAck(Message message) {
+        return message.getSequence() == message.getBatch().getBatchSize();
+    }
+
+    private void ack(ChannelHandlerContext ctx, Message message) {
+        writeAck(ctx, message.getBatch().getProtocol(), message.getSequence());
+    }
+
+    private void writeAck(ChannelHandlerContext ctx, byte protocol, int sequence) {
+        ByteBuf buffer = ctx.alloc().buffer(6);
+        buffer.writeByte(protocol);
+        buffer.writeByte('A');
+        buffer.writeInt(sequence);
+        ctx.writeAndFlush(buffer);
```

Comment:
From the doc, I pretty sure I read that buffer will be release when the write occur?